lib/tlIO/FFmpegWrite.cpp

```diff
@@ -42,198 +176,582 @@ namespace tl
 
-            //av_dump_format(p.avFormatContext, 0, p.fileName.c_str(), 1);
+                p.sampleRate =
+                    selectSampleRate(avCodec, info.audio.sampleRate);
+                if (p.sampleRate == 0)
+                    throw std::runtime_error(
+                        string::Format(
+                            "{0}: Could not select sample rate")
+                            .arg(p.fileName));
+                    
+                char buf[256];
+                av_channel_layout_describe(
+                    &p.avAudioCodecContext->ch_layout, buf, 256);
+                    
+                if (p.sampleRate != info.audio.sampleRate || resample)
+                {
+                    const audio::Info& input = info.audio;
+                    audio::Info output(info.audio.channelCount,
+                                       toAudioType(p.avAudioCodecContext->sample_fmt),
+                                       p.sampleRate);
+                    p.resample = audio::AudioResample::create(input, output);
+                    
+                    if (auto logSystem = _logSystem.lock())
+                    {
+                        logSystem->print(
+                            "tl::io::ffmpeg::Plugin::Write",
+                            string::Format(
+                                "Resample from layout {0}, {1} channels, type "
+                                "{2}, sample rate {3} to layout {4}, {5} "
+                                "channels, type {6}, sample rate {7}.")
+                            .arg(buf)
+                            .arg(input.channelCount)
+                            .arg(input.dataType)
+                            .arg(input.sampleRate)
+                            .arg(buf)
+                            .arg(output.channelCount)
+                            .arg(output.dataType)
+                            .arg(output.sampleRate));
+                    }
+                }
+                else
+                {
+                    const audio::Info& input = info.audio;
+                    if (auto logSystem = _logSystem.lock())
+                    {
+                        logSystem->print(
+                            "tl::io::ffmpeg::Plugin::Write",
+                            string::Format(
+                                "Save from layout {0}, {1} channels, type "
+                                "{2}, sample rate {3}.")
+                            .arg(buf)
+                            .arg(input.channelCount)
+                            .arg(input.dataType)
+                            .arg(input.sampleRate));
+                    }
+                }
 
-            r = avio_open(&p.avFormatContext->pb, p.fileName.c_str(), AVIO_FLAG_WRITE);
-            if (r < 0)
-            {
-                throw std::runtime_error(string::Format("{0}: {1}").arg(p.fileName).arg(getErrorLabel(r)));
-            }
+                p.avAudioCodecContext->bit_rate = 69000;
```

Comment:
I like to try and put hard-coded numbers in an anonymous namespace at the top of the file for visibility:
```
namespace
{
    const size_t bitRateDefault = 69000;
}
```